core/server/policies.go

```diff
@@ -0,0 +1,260 @@
+	} else {
+		clustersClient, err = cs.clustersManager.GetImpersonatedClient(ctx, auth.Principal(ctx))
+	}
+
+	if err != nil {
+		if merr, ok := err.(*multierror.Error); ok {
+			for _, err := range merr.Errors {
+				if cerr, ok := err.(*clustersmngr.ClientError); ok {
+					respErrors = append(respErrors, &pb.ListError{ClusterName: cerr.ClusterName, Message: cerr.Error()})
+				}
+			}
+		} else {
+			return nil, fmt.Errorf("unexpected error while getting clusters client, error: %w", err)
+		}
+	}
+
+	opts := []client.ListOption{}
+	if m.Pagination != nil {
+		opts = append(opts, client.Limit(m.Pagination.PageSize))
+		opts = append(opts, client.Continue(m.Pagination.PageToken))
+	}
+
+	var continueToken string
+	var listsV2beta2 map[string][]client.ObjectList
+
+	if m.ClusterName == "" {
+		clistV2beta2 := clustersmngr.NewClusteredList(func() client.ObjectList {
+			return &pacv2beta2.PolicyList{}
+		})
+
+		var errsV2beta2 clustersmngr.ClusteredListError
+
+		if err := clustersClient.ClusteredList(ctx, clistV2beta2, false, opts...); err != nil {
+			if !errors.As(err, &errsV2beta2) {
+				return nil, fmt.Errorf("error while listing v2beta2 policies: %w", err)
+			}
+		}
+		for _, e := range errsV2beta2.Errors {
+			if !strings.Contains(e.Err.Error(), "no matches for kind \"Policy\"") {
+				respErrors = append(respErrors, &pb.ListError{ClusterName: e.Cluster, Message: e.Err.Error()})
+			}
+		}
+
+		continueToken = clistV2beta2.GetContinue()
+		listsV2beta2 = clistV2beta2.Lists()
+	} else {
+		listV2beta2 := &pacv2beta2.PolicyList{}
+
+		if err := clustersClient.List(ctx, m.ClusterName, listV2beta2, opts...); err != nil {
+			return nil, fmt.Errorf("error while listing policies for cluster %s: %w", m.ClusterName, err)
+		}
+
+		continueToken = listV2beta2.GetContinue()
+		listsV2beta2 = map[string][]client.ObjectList{m.ClusterName: {listV2beta2}}
+	}
+
+	var policies []*pb.Policy
+	for clusterName, lists := range listsV2beta2 {
+		for _, l := range lists {
+			list, ok := l.(*pacv2beta2.PolicyList)
+			if !ok {
+				respErrors = append(respErrors, &pb.ListError{ClusterName: clusterName, Message: fmt.Sprintf("unexpected list type %T", l)})
+				continue
+			}
+			for i := range list.Items {
+				policy, err := policyToPolicyRespone(list.Items[i], clusterName, false)
+				if err != nil {
+					return nil, err
+				}
+
+				policies = append(policies, policy)
+			}
+		}
+	}
+
+	return &pb.ListPoliciesResponse{
+		Policies:      policies,
+		Total:         int32(len(policies)),
+		NextPageToken: continueToken,
+		Errors:        respErrors,
+	}, nil
+}
+
+func (cs *coreServer) GetPolicy(ctx context.Context, m *pb.GetPolicyRequest) (*pb.GetPolicyResponse, error) {
+	var clustersClient clustersmngr.Client
+	var err error
+
+	if m.ClusterName != "" {
+		clustersClient, err = cs.clustersManager.GetImpersonatedClientForCluster(ctx, auth.Principal(ctx), m.ClusterName)
+	} else {
+		m.ClusterName = DefaultCluster
```

Comment:
why dont you just set the clusterName if it's empty and then use the same client for both cases.